File: src/test/java/com/gitlab/jactor/persistence/repository/GuestBookRepositoryTest.java
package com.gitlab.jactor.persistence.repository;

import com.gitlab.jactor.persistence.JactorPersistence;
import com.gitlab.jactor.persistence.entity.address.AddressEntity;
import com.gitlab.jactor.persistence.entity.guestbook.GuestBookEntity;
import com.gitlab.jactor.persistence.entity.person.PersonEntity;
import com.gitlab.jactor.persistence.entity.user.UserEntity;
import com.gitlab.jactor.persistence.fields.FieldValue;
import com.gitlab.jactor.persistence.fields.RequiredFieldsExtension;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.api.extension.RegisterExtension;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

import static com.gitlab.jactor.persistence.entity.address.AddressEntity.anAddress;
import static com.gitlab.jactor.persistence.entity.guestbook.GuestBookEntity.aGuestBook;
import static com.gitlab.jactor.persistence.entity.person.PersonEntity.aPerson;
import static com.gitlab.jactor.persistence.entity.user.UserEntity.aUser;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

@ExtendWith(SpringExtension.class)
@SpringBootTest(classes = {JactorPersistence.class})
@Transactional
@DisplayName("A GuestBookRepository")
class GuestBookRepositoryTest {

    @RegisterExtension RequiredFieldsExtension requiredFieldsExtension = new RequiredFieldsExtension(Map.of(
            UserEntity.class, List.of(
                    new FieldValue("username", () -> "unique@" + LocalDateTime.now()),
                    new FieldValue("personEntity", () -> aPerson().build())
            ), PersonEntity.class, List.of(
                    new FieldValue("addressEntity", () -> anAddress().build()),
                    new FieldValue("surname", "sure, man")
            ), AddressEntity.class, List.of(
                    new FieldValue("addressLine1", "Test Boulevard 1"),
                    new FieldValue("zipCode", 1001),
                    new FieldValue("city", "Testing")
            )
    ));

    private @Autowired GuestBookRepository guestBookRepository;
    private @Autowired EntityManager entityManager;

    @DisplayName("should write then read guest book")
    @Test void shouldWriteThenReadGuestBook() {
        GuestBookEntity guestBookEntityToSave = aGuestBook()
                .withTitle("home sweet home")
                .with(aUser())
                .build();

        guestBookRepository.save(guestBookEntityToSave);
        entityManager.flush();
        entityManager.clear();

        GuestBookEntity guestBookEntity = guestBookRepository.findById(guestBookEntityToSave.getId()).orElseThrow(this::guestBookNotFound);

        assertAll(
                () -> assertThat(guestBookEntity.getTitle()).as("title").isEqualTo("home sweet home"),
                () -> assertThat(guestBookEntity.getUser()).as("user").isNotNull()
        );
    }

    @DisplayName("should write then update and read guest book")
    @Test void shouldWriteThenUpdateAndReadGuestBook() {
        GuestBookEntity guestBookEntityToSave = aGuestBook()
                .withTitle("home sweet home")
                .with(aUser())
                .build();

        guestBookRepository.save(guestBookEntityToSave);
        entityManager.flush();
        entityManager.clear();

        GuestBookEntity guestBookEntityToUpdate = guestBookRepository.findById(guestBookEntityToSave.getId()).orElseThrow(this::guestBookNotFound);

        guestBookEntityToUpdate.setTitle("5000 thousands miles away from home");

        guestBookRepository.save(guestBookEntityToUpdate);
        entityManager.flush();
        entityManager.clear();

        GuestBookEntity guestBookEntity = guestBookRepository.findById(guestBookEntityToSave.getId()).orElseThrow(this::guestBookNotFound);

        assertThat(guestBookEntity.getTitle()).isEqualTo("5000 thousands miles away from home");
    }

    private AssertionError guestBookNotFound() {
        return new AssertionError("Guest book not found");
    }
}

File: src/test/kotlin/com/github/jactor/persistence/guestbook/GuestBookRepositoryTest.kt
package com.github.jactor.persistence.guestbook

import java.util.UUID
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.fail
import org.springframework.beans.factory.annotation.Autowired
import com.github.jactor.persistence.AbstractSpringBootNoDirtyContextTest
import com.github.jactor.persistence.AddressModel
import com.github.jactor.persistence.common.PersistentModel
import com.github.jactor.persistence.person.PersonModel
import com.github.jactor.persistence.user.UserModel
import com.github.jactor.persistence.AddressBuilder
import com.github.jactor.persistence.person.PersonBuilder
import com.github.jactor.persistence.user.UserBuilder
import com.github.jactor.persistence.user.UserRepository
import assertk.assertAll
import assertk.assertThat
import assertk.assertions.isEqualTo
import assertk.assertions.isNotNull

internal class GuestBookRepositoryTest : AbstractSpringBootNoDirtyContextTest() {
    @Autowired
    private lateinit var guestBookRepository: GuestBookRepository

    @Autowired
    private lateinit var userRepository: UserRepository

    @Test
    fun `should write then read guest book`() {
        val addressDto = AddressBuilder
            .new(
                addressModel = AddressModel(
                    zipCode = "1001",
                    addressLine1 = "Test Boulevard 1",
                    city = "Testington"
                )
            )
            .addressModel

        val personDto = PersonModel(
            persistentModel = PersistentModel(id = UUID.randomUUID()), address = addressDto, surname = "AA"
        )

        val userDto = UserModel(
            PersistentModel(id = UUID.randomUUID()),
            personInternal = personDto,
            emailAddress = "casuel@tantooine.com",
            username = "causual"
        )

        val userEntity = userRepository.save(
            UserBuilder.new(userDto = userDto).build()
        )

        userEntity.setGuestBook(
            GuestBookBuilder.new(
                GuestBookModel(
                    entries = emptySet(),
                    title = "home sweet home",
                    user = userEntity.toModel()
                )
            ).buildGuestBookEntity()
        )

        flush { }

        val guestBookEntity = guestBookRepository.findByUser(userEntity)

        assertAll {
            assertThat(guestBookEntity?.title).isEqualTo("home sweet home")
            assertThat(guestBookEntity?.user).isNotNull()
        }
    }

    @Test
    fun `should write then update and read guest book`() {
        val addressDto = AddressBuilder.new(
            addressModel = AddressModel(
                zipCode = "1001", addressLine1 = "Test Boulevard 1", city = "Testington"
            )
        ).addressModel

        val personDto = PersonBuilder.new(PersonModel(address = addressDto, surname = "AA")).personModel
        val userDto = UserBuilder.unchanged(
            userModel = UserModel(
                persistentModel = PersistentModel(),
                personInternal = personDto,
                emailAddress = "casuel@tantooine.com",
                username = "causual"
            )
        ).userDto

        val userEntity = userRepository.save(UserBuilder.new(userDto).build())

        userEntity.setGuestBook(
            GuestBookBuilder.new(
                GuestBookModel(
                    entries = emptySet(),
                    title = "home sweet home",
                    user = userEntity.toModel()
                )
            ).buildGuestBookEntity()
        )

        flush { guestBookRepository.save(userEntity.guestBook ?: fail(message = "User missing guest book")) }

        val guestBookEntityToUpdate = guestBookRepository.findByUser(userEntity)

        guestBookEntityToUpdate!!.title = "5000 thousands miles away from home"

        flush { guestBookRepository.save(guestBookEntityToUpdate) }

        val guestBookEntity = guestBookRepository.findByUser(userEntity)

        assertThat(guestBookEntity!!.title).isEqualTo("5000 thousands miles away from home")
    }
}
